addons/sale/models/sale_order.py

```diff
@@ -788,6 +788,14 @@ def _unlink_except_draft_or_cancel(self):
                     "You can not delete a sent quotation or a confirmed sales order."
                     " You must first cancel it."))
 
+    def write(self, vals):
+        for order in self:
+            old_partner_id = order.partner_id.id
+            new_partner_id = vals.get('partner_id', old_partner_id)
```

Comment:
You could maybe first check if `partner_id` is in `vals` and skip the loop over `self` if it is not.